Cloning an index using 7.11 causes RecoveryFailedException

**Elasticsearch version** (`bin/elasticsearch --version`): 7.11.0

**Plugins installed**: []

**JVM version** (`java -version`): Cloud

**OS version** (`uname -a` if on a Unix-like system): Cloud

**Description of the problem including expected versus actual behavior**:

I created a deployment in the QA env on the cloud. I then created an index and filled that index with some data (using esrally). I then tried to `_clone` the index. The newly created index went immediately into a red state and never recovers. 

**Steps to reproduce**:

 1. Create cluster with 7.11
 2. Create index and load with some data
 3. Set the index to read only mode
 4. Call `_clone` api (https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-clone-index.html)

**Provide logs (if relevant)**:
Calling `_cluster/allocation/explain` yielded:

```
{
  "can_allocate": "yes",
  "index": "pmc_c1",
  "target_node": {
    "attributes": {
      "server_name": "instance-0000000000.563fa5ba6e6d4b0795c587d43f2ee605",
      "availability_zone": "eu-west-1a",
      "transform.node": "true",
      "region": "eu-west-1",
      "instance_configuration": "aws.data.highio.i3",
      "xpack.installed": "true",
      "logical_availability_zone": "zone-0"
    },
    "transport_address": "172.23.177.56:19897",
    "id": "J132sguIQmK9uWOWPX6ItA",
    "name": "instance-0000000000"
  },
  "node_allocation_decisions": [
    {
      "node_decision": "yes",
      "transport_address": "172.23.177.56:19897",
      "weight_ranking": 1,
      "node_name": "instance-0000000000",
      "node_id": "J132sguIQmK9uWOWPX6ItA",
      "node_attributes": {
        "server_name": "instance-0000000000.563fa5ba6e6d4b0795c587d43f2ee605",
        "availability_zone": "eu-west-1a",
        "transform.node": "true",
        "region": "eu-west-1",
        "instance_configuration": "aws.data.highio.i3",
        "xpack.installed": "true",
        "logical_availability_zone": "zone-0"
      }
    },
    {
      "node_decision": "no",
      "transport_address": "172.23.178.44:19934",
      "weight_ranking": 2,
      "node_name": "instance-0000000001",
      "node_id": "Tlk4BLZkTHemXDYALai4-Q",
      "deciders": [
        {
          "decision": "NO",
          "explanation": "source primary is allocated on another node",
          "decider": "resize"
        }
      ],
      "node_attributes": {
        "server_name": "instance-0000000001.563fa5ba6e6d4b0795c587d43f2ee605",
        "availability_zone": "eu-west-1b",
        "transform.node": "true",
        "region": "eu-west-1",
        "instance_configuration": "aws.data.highio.i3",
        "xpack.installed": "true",
        "logical_availability_zone": "zone-1"
      }
    }
  ],
  "current_state": "unassigned",
  "shard": 3,
  "primary": true,
  "allocate_explanation": "can allocate the shard",
  "unassigned_info": {
    "last_allocation_status": "no",
    "reason": "ALLOCATION_FAILED",
    "failed_allocation_attempts": 5,
    "at": "2021-01-07T14:14:36.025Z",
    "details": "failed shard on node [J132sguIQmK9uWOWPX6ItA]: failed recovery, failure RecoveryFailedException[[pmc_c1][3]: Recovery failed on {instance-0000000000}{J132sguIQmK9uWOWPX6ItA}{3QGTtbcqTEas8xtG3IuT1A}{172.23.177.56}{172.23.177.56:19897}{cdhimrstw}{logical_availability_zone=zone-0, server_name=instance-0000000000.563fa5ba6e6d4b0795c587d43f2ee605, availability_zone=eu-west-1a, xpack.installed=true, instance_configuration=aws.data.highio.i3, transform.node=true, region=eu-west-1}]; nested: IndexShardRecoveryException[failed recovery]; nested: IllegalArgumentException; "
  }
}
```

The stack trace in the logs shows:
```
[instance-0000000000] failing shard [failed shard, shard [geonames_clone2][1], node[F9F6lSs8RZmO7yikMdiwBw], [P], recovery_source[local shards recovery], s[INITIALIZING], a[id=OMTGoVavRSqbJczJrvXWgw], unassigned_info[[reason=ALLOCATION_FAILED], at[2021-01-06T22:20:26.043Z], failed_attempts[1], failed_nodes[[F9F6lSs8RZmO7yikMdiwBw]], delayed=false, details[failed shard on node [F9F6lSs8RZmO7yikMdiwBw]: failed recovery, failure RecoveryFailedException[[geonames_clone2][1]: Recovery failed on {instance-0000000001}{F9F6lSs8RZmO7yikMdiwBw}{QTylI4zWShaiYujng0f3rQ}{172.23.178.44}{172.23.178.44:19440}{cdhimrstw}{logical_availability_zone=zone-1, server_name=instance-0000000001.d072b1b4f7aa4abb8487bee27e1bd769, availability_zone=eu-west-1b, xpack.installed=true, instance_configuration=aws.data.highio.i3, transform.node=true, region=eu-west-1}]; nested: IndexShardRecoveryException[failed recovery]; nested: IllegalArgumentException; ], allocation_status[deciders_no]], expected_shard_size[498523189], message [failed recovery], failure [RecoveryFailedException[[geonames_clone2][1]: Recovery failed on {instance-0000000001}{F9F6lSs8RZmO7yikMdiwBw}{QTylI4zWShaiYujng0f3rQ}{172.23.178.44}{172.23.178.44:19440}{cdhimrstw}{logical_availability_zone=zone-1, server_name=instance-0000000001.d072b1b4f7aa4abb8487bee27e1bd769, availability_zone=eu-west-1b, xpack.installed=true, instance_configuration=aws.data.highio.i3, transform.node=true, region=eu-west-1}]; nested: IndexShardRecoveryException[failed recovery]; nested: IllegalArgumentException; ], markAsStale [true]]
org.elasticsearch.indices.recovery.RecoveryFailedException: [geonames_clone2][1]: Recovery failed on {instance-0000000001}{F9F6lSs8RZmO7yikMdiwBw}{QTylI4zWShaiYujng0f3rQ}{172.23.178.44}{172.23.178.44:19440}{cdhimrstw}{logical_availability_zone=zone-1, server_name=instance-0000000001.d072b1b4f7aa4abb8487bee27e1bd769, availability_zone=eu-west-1b, xpack.installed=true, instance_configuration=aws.data.highio.i3, transform.node=true, region=eu-west-1}
	at org.elasticsearch.index.shard.IndexShard.lambda$executeRecovery$21(IndexShard.java:2827) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:136) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.action.ActionListener$5.onFailure(ActionListener.java:307) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoveryListener$6(StoreRecovery.java:363) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:136) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:343) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromLocalShards(StoreRecovery.java:129) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.index.shard.IndexShard.recoverFromLocalShards(IndexShard.java:1979) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$19(IndexShard.java:2800) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:73) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:737) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.11.0.jar:7.11.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
	at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: org.elasticsearch.index.shard.IndexShardRecoveryException: failed recovery
	... 12 more
Caused by: java.lang.IllegalArgumentException
	at sun.nio.fs.UnixPath.toUnixPath(UnixPath.java:198) ~[?:?]
	at sun.nio.fs.UnixFileSystemProvider.createLink(UnixFileSystemProvider.java:467) ~[?:?]
	at org.elasticsearch.fs.quotaaware.QuotaAwareFileSystemProvider.createLink(QuotaAwareFileSystemProvider.java:411) ~[?:7.11.0]
	at java.nio.file.Files.createLink(Files.java:1106) ~[?:?]
	at org.apache.lucene.store.HardlinkCopyDirectoryWrapper.lambda$copyFrom$0(HardlinkCopyDirectoryWrapper.java:67) ~[lucene-misc-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:35:28]
	at java.security.AccessController.doPrivileged(AccessController.java:312) ~[?:?]
	at org.apache.lucene.store.HardlinkCopyDirectoryWrapper.copyFrom(HardlinkCopyDirectoryWrapper.java:65) ~[lucene-misc-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:35:28]
	at org.elasticsearch.index.shard.StoreRecovery$StatsDirectoryWrapper.copyFrom(StoreRecovery.java:216) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.apache.lucene.store.LockValidatingDirectoryWrapper.copyFrom(LockValidatingDirectoryWrapper.java:50) ~[lucene-core-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:35:28]
	at org.apache.lucene.index.IndexWriter.copySegmentAsIs(IndexWriter.java:3244) ~[lucene-core-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:35:28]
	at org.apache.lucene.index.IndexWriter.addIndexes(IndexWriter.java:2987) ~[lucene-core-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:35:28]
	at org.elasticsearch.index.shard.StoreRecovery.addIndices(StoreRecovery.java:177) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromLocalShards$3(StoreRecovery.java:137) ~[elasticsearch-7.11.0.jar:7.11.0]
	at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:340) ~[elasticsearch-7.11.0.jar:7.11.0]
	... 9 more
```

--- a/x-pack/quota-aware-fs/src/main/java/org/elasticsearch/fs/quotaaware/QuotaAwareFileSystemProvider.java
+++ b/x-pack/quota-aware-fs/src/main/java/org/elasticsearch/fs/quotaaware/QuotaAwareFileSystemProvider.java
@@ -408,12 +408,12 @@ public class QuotaAwareFileSystemProvider extends FileSystemProvider implements 
 
     @Override
     public void createLink(Path link, Path existing) throws IOException {
-        delegate.createLink(link, existing);
+        delegate.createLink(QuotaAwarePath.unwrap(link), QuotaAwarePath.unwrap(existing));
     }
 
     @Override
     public void createSymbolicLink(Path link, Path target, FileAttribute<?>... attrs) throws IOException {
-        delegate.createSymbolicLink(link, target, attrs);
+        delegate.createSymbolicLink(QuotaAwarePath.unwrap(link), QuotaAwarePath.unwrap(target), attrs);
     }
 
     void purge(FileSystem delegateFileSystem) {

--- a/x-pack/quota-aware-fs/src/test/java/org/elasticsearch/fs/quotaaware/QuotaAwareFileSystemProviderTests.java
+++ b/x-pack/quota-aware-fs/src/test/java/org/elasticsearch/fs/quotaaware/QuotaAwareFileSystemProviderTests.java
@@ -20,8 +20,10 @@ import java.net.URISyntaxException;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.FileStore;
 import java.nio.file.FileSystems;
+import java.nio.file.Files;
 import java.nio.file.NoSuchFileException;
 import java.nio.file.Path;
+import java.nio.file.attribute.FileAttribute;
 import java.nio.file.spi.FileSystemProvider;
 import java.security.PrivilegedActionException;
 import java.util.Properties;
@@ -275,6 +277,48 @@ public class QuotaAwareFileSystemProviderTests extends LuceneTestCase {
         try (QuotaAwareFileSystemProvider provider = new QuotaAwareFileSystemProvider(cyclicProvider, quotaFile.toUri())) {
             cyclicProvider.cyclicReference = provider;
             assertNotNull(provider.getPath(new URI("file:///")));
+        }
+    }
+
+    /**
+     * If the implementation of {@link QuotaAwareFileSystemProvider#createLink(Path, Path)}
+     * doesn't unwrap its {@link Path} arguments, it causes a runtime exception, so exercise
+     * this method to check that it unwraps correctly.
+     */
+    public void testCreateLinkUnwrapsPaths() throws Exception {
+        final Path tempDir = createTempDir();
+        Path quotaFile = tempDir.resolve("quota.properties");
+        FileSystemProvider systemProvider = quotaFile.getFileSystem().provider();
+        writeQuota(500, 200, systemProvider, quotaFile);
+
+        try (QuotaAwareFileSystemProvider provider = new QuotaAwareFileSystemProvider(systemProvider, quotaFile.toUri())) {
+            final Path quotaFilePath = provider.getPath(tempDir.resolve("path1.txt").toUri());
+            final Path quotaLinkPath = provider.getPath(tempDir.resolve("path2.txt").toUri());
+
+            Files.writeString(quotaFilePath, "some text");
+
+            provider.createLink(quotaLinkPath, quotaFilePath);
+        }
+    }
+
+    /**
+     * If the implementation of {@link QuotaAwareFileSystemProvider#createSymbolicLink(Path, Path, FileAttribute[])}
+     * doesn't unwrap its {@link Path} arguments, it causes a runtime exception, so exercise
+     * this method to check that it unwraps correctly.
+     */
+    public void testCreateSymbolicLinkUnwrapsPaths() throws Exception {
+        final Path tempDir = createTempDir();
+        Path quotaFile = tempDir.resolve("quota.properties");
+        FileSystemProvider systemProvider = quotaFile.getFileSystem().provider();
+        writeQuota(500, 200, systemProvider, quotaFile);
+
+        try (QuotaAwareFileSystemProvider provider = new QuotaAwareFileSystemProvider(systemProvider, quotaFile.toUri())) {
+            final Path quotaFilePath = provider.getPath(tempDir.resolve("path1.txt").toUri());
+            final Path quotaLinkPath = provider.getPath(tempDir.resolve("path2.txt").toUri());
+
+            Files.writeString(quotaFilePath, "some text");
+
+            provider.createSymbolicLink(quotaLinkPath, quotaFilePath);
         }
     }
 

